text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,621 @@
+stable way to name items across compilations (and across crates).  The
+compiler currently uses something called a `DefId` to identify each
+item. However, these ids today are based on a node-id, which is just
+an index into the HIR and hence will change whenever *anything*
+preceding it in the HIR changes. We need to make the `DefId` for an
+item independent of changes to other items.
+
+Conceptually, the idea is to change `DefId` into the pair of a crate
+and a path:
+
+```
+CRATE = <crate identifier>
+PATH = Crate(ID)
+     | PATH :: Mod(ID)
+     | PATH :: Item(ID)
+     | PATH :: TypeParameter(ID)
+     | PATH :: LifetimeParameter(ID)
+     | PATH :: Member(ID)
+     | PATH :: Impl
+     | ...
+```
+
+However, rather than actually store the path in the compiler, we will
+instead intern the paths in the `CStore`, and the `DefId` will simply
+store an integer. So effectively the `node` field of `DefId`, which
+currently indexes into the HIR of the appropriate crate, becomes an
+index into the crate's list of paths.
+
+For the most part, these paths match up with user's intutions. So a
+struct `Foo` declared in a module `bar` would just have a path like
+`bar::Foo`. However, the paths are also able to express things for
+which there is no syntax, such as an item declared within a function
+body.
+
+### Disambiguation
+
+For the most part, paths should naturally be unique. However, there
+are some cases where a single parent may have multiple children with
+the same path. One case would be erroneous programs, where there are
+(e.g.) two structs declared with the same name in the same
+module. Another is that some items, such as impls, do not have a name,
+and hence we cannot easily distinguish them. Finally, it is possible
+to declare multiple functions with the same name within function bodies:
+
+```rust
+fn foo() {
+    {
+        fn bar() { }
+    }
+
+    {
+        fn bar() { }
+    }
+}
+```
+
+All of these cases are handled by a simple *disambiguation* mechanism.
+The idea is that we will assign a path to each item as we traverse the
+HIR. If we find that a single parent has two children with the same
+name, such as two impls, then we simply assign them unique integers in
+the order that they appear in the program text. For example, the
+following program would use the paths shown:
+
+```rust
+mod foo {               // Path: <root>::foo
+    pub struct Type { } // Path: <root>::foo::Type
+    impl Type {         // Path: <root>::foo::<impl1>
+        fn bar() {..}   // Path: <root>::foo::<impl1>::bar
+    }
+    impl Type { }       // Path: <root>::foo::<impl2>
+}
+```
+
+Note that the impls were arbitarily assigned indices based on the order
+in which they appear. This does mean that reordering impls may cause
+spurious recompilations. We can try to mitigate this somewhat by making the
+path entry for an impl include some sort of hash for its header or its contents,
+but that will be something we can add later.
```

Comment:
I can't think of any concrete cases, but it seems unlikely that this would be the only spurious recompilation. If you know of others, it might be good to list those somewhere. Maybe even as a staged list (corresponding to your list above?), stating that when we cache object files, X and Y cause unneeded recompilation, but after that only X does.